src/hll/sketch.rs

```diff
@@ -0,0 +1,422 @@
+                if should_promote {
+                    self.mode = if self.lg_config_k < 8 {
+                        promote_container_to_array(list.container(), *hll_type, self.lg_config_k)
+                    } else {
+                        promote_container_to_set(list.container(), *hll_type)
+                    }
+                }
+            }
+            Mode::Set { set, hll_type } => {
+                set.update(coupon);
+                let should_promote = RESIZE_DENOM as usize * set.container().len()
+                    > RESIZE_NUMER as usize * set.container().capacity();
+                if should_promote {
+                    self.mode = if set.container().lg_size() == self.lg_config_k as usize - 3 {
+                        promote_container_to_array(set.container(), *hll_type, self.lg_config_k)
+                    } else {
+                        grow_set(set, *hll_type)
+                    }
+                }
+            }
+            Mode::Array4(arr) => arr.update(coupon),
+            Mode::Array6(arr) => arr.update(coupon),
+            Mode::Array8(arr) => arr.update(coupon),
+        }
+    }
+
+    /// Get the current cardinality estimate
+    pub fn estimate(&self) -> f64 {
+        match &self.mode {
+            Mode::List { list, .. } => list.container().estimate(),
+            Mode::Set { set, .. } => set.container().estimate(),
+            Mode::Array4(arr) => arr.estimate(),
+            Mode::Array6(arr) => arr.estimate(),
+            Mode::Array8(arr) => arr.estimate(),
+        }
+    }
+
+    /// Get upper bound for cardinality estimate
+    ///
+    /// Returns the upper confidence bound for the cardinality estimate based on
+    /// the number of standard deviations requested.
+    pub fn upper_bound(&self, num_std_dev: u8) -> f64 {
+        match &self.mode {
+            Mode::List { list, .. } => list.container().upper_bound(num_std_dev),
+            Mode::Set { set, .. } => set.container().upper_bound(num_std_dev),
+            Mode::Array4(arr) => arr.upper_bound(num_std_dev),
+            Mode::Array6(arr) => arr.upper_bound(num_std_dev),
+            Mode::Array8(arr) => arr.upper_bound(num_std_dev),
+        }
+    }
+
+    /// Get lower bound for cardinality estimate
+    ///
+    /// Returns the lower confidence bound for the cardinality estimate based on
+    /// the number of standard deviations requested.
+    pub fn lower_bound(&self, num_std_dev: u8) -> f64 {
+        match &self.mode {
+            Mode::List { list, .. } => list.container().lower_bound(num_std_dev),
+            Mode::Set { set, .. } => set.container().lower_bound(num_std_dev),
+            Mode::Array4(arr) => arr.lower_bound(num_std_dev),
+            Mode::Array6(arr) => arr.lower_bound(num_std_dev),
+            Mode::Array8(arr) => arr.lower_bound(num_std_dev),
+        }
+    }
+
+    pub fn deserialize(bytes: &[u8]) -> io::Result<HllSketch> {
```

Comment:
For all `serialize`/`deserialize` methods here, since we don't do IO anyway, I suggest we define DataSketches serde error.

We may further optionally integrate with the [`serde`](https://serde.rs/impl-serialize.html) crate but ser/de sketches should work without `serde` dep. (Its usage requires a serializer/deserializer that involves other concepts beyond sketches from/to bytes)